src/app/admin/admin-campaign-new/admin-campaign-new.component.ts

```diff
@@ -0,0 +1,295 @@
+
+  onDocumentFilesAdded(newDocuments: File[]) {
+    const documents = this.createForm2.value.documents.concat(newDocuments)
+    this.createForm2.controls.documents.setValue(documents)
+  }
+
+  removeDocumentFile(index: number) {
+    this.createForm2.value.documents.splice(index, 1)
+    this.createForm2.markAsDirty()
+  }
+
+  newsUrlsControls() {
+    return this.newsUrls.controls as FormControl[]
+  }
+
+  addNewsUrl() {
+    this.newsUrls.push(this.fb.control('', [Validators.required, Validators.pattern('[^\\s]*')]))
+  }
+
+  removeNewsUrl(index: number) {
+    this.newsUrls.removeAt(index)
+    this.newsUrls.markAsDirty()
+  }
+
+  private tokenPercentage(pricePerToken: number, totalValue: number) {
+    if (pricePerToken === 0) {
+      return 0
+    }
+
+    const numOfTokensToSell = totalValue / pricePerToken
+    if (numOfTokensToSell === 0) {
+      return 0
+    }
+
+    const totalTokens = this.stablecoinService.format(this.assetData.asset.initialTokenSupply, 18)
+    return numOfTokensToSell / totalTokens
+  }
+
+  private validMonetaryValues(formGroup: FormGroup): ValidationErrors | null {
+    const hardCap = formGroup.value.hardCap
+    if (hardCap <= 0) {
+      return {invalidHardCap: true}
+    }
+
+    const softCap = formGroup.value.softCap
+    if (softCap <= 0 || softCap > hardCap) {
+      return {invalidSoftCap: true}
+    }
+
+    const hardCapTokensPercentage = formGroup.value.hardCapTokensPercentage
+    if (hardCapTokensPercentage <= 0 || hardCapTokensPercentage > this.maxTokensPercentage) {
+      return {invalidHardCapTokensMaxPercentage: true}
+    }
+
+    if (formGroup.value.tokenPrice <= 0) {
+      return {invalidTokenPrice: true}
+    }
+
+    if (formGroup.value.hasMinAndMaxInvestment) {
+      const minInvestment = formGroup.value.minInvestment
+      const maxInvestment = formGroup.value.maxInvestment
+      if (minInvestment <= 0 || minInvestment > maxInvestment) {
+        return {invalidMinInvestment: true}
+      }
+
+      if (maxInvestment <= 0 || maxInvestment > formGroup.value.hardCap) {
+        return {invalidMaxInvestment: true}
+      }
+    }
+
+    return null
+  }
+
+  private static validReturnFromTo(formGroup: FormGroup): ValidationErrors | null {
+    if (formGroup.value.isReturningProfitsToInvestors) {
+      const returnFrom = formGroup.value.returnFrom
+      if (returnFrom <= 0 || returnFrom > 1) {
+        return {invalidReturnFrom: true}
+      }
+
+      if (!formGroup.value.isReturnValueFixed) {
+        const returnTo = formGroup.value.returnTo
+        if (returnTo <= 0 || returnTo > 1) {
+          return {invalidReturnTo: true}
+        }
+
+        if (returnFrom >= returnTo) {
+          return {invalidReturnFromToRange: true}
+        }
+      }
+    }
+
+    return null
+  }
+
+  private static validDateRange(formGroup: FormGroup): ValidationErrors | null {
+    if (formGroup.value.startDate >= formGroup.value.endDate) {
+      return {invalidDateRange: true}
+    }
```

Comment:
Isn't `startDate == endDate` a valid input? Theoretically, it is possible to have a one-day period for a project. 

![image](https://user-images.githubusercontent.com/13491551/133056799-8bc00315-b65c-4b60-ba3a-36f3bbe764af.png)
😆

Reply:
True, I will fix that 😄